README.md
# flutter_highlight_text
Flutter文本指定文字或者指定分割符高亮,高亮文字点击之后回调点击内容

### 使用方法
```
class TextPage extends StatefulWidget {
  @override
  State<StatefulWidget> createState() => _TextPageState();
}

class _TextPageState extends State<TextPage> {
  String text =
      '可能说起 Flutter 绘制，大家第一反应就是用 `CustomPaint` 组件，自定义 `CustomPainter` 对象来画。Flutter 中所有可以看得到的组件，比如 Text、Image、Switch、Slider 等等，追其根源都是`画出来`的，但通过查看源码可以发现，Flutter 中绝大多数组件并不是使用 `CustomPaint` 组件来画的，其实 `CustomPaint` 组件是对框架底层绘制的一层封装。这个系列便是对 Flutter 绘制的探索，通过`测试`、`调试`及`源码分析`来给出一些在绘制时`被忽略`或`从未知晓`的东西，而有些要点如果被忽略，就很可能出现问题。';

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('demo'),
      ),
      body: Container(
          color: Colors.white,
          child: CustomScrollView(
            slivers: [getText()],
          )),
    );
  }

  SliverToBoxAdapter getText() {
    return SliverToBoxAdapter(
      child: HighLightText(
        style: TextStyle(fontSize: 14, color: Colors.black),
        highLightTextStyle:
            TextStyle(color: Colors.red, fontSize: 18, fontWeight: FontWeight.bold),
        text: text,
        highLightWords: ['Flutter', '可'],
        highLightType: HighLightType.keyWords,
        //以分隔符分割
        //splitString: '`',
        //highLightType: HighLightType.splitCharacter,
        wordClickAction: (word) {
          print("=====$word");
        },
      ),
    );
  }
}



CHANGELOG.md
## 0.0.4

将kmp_util调整为String的extension


lib/src/util/kmp_util.dart
import 'package:flutter/cupertino.dart';

extension ExtensionString on String {
  _calculateNext(String string, List<int> next) {
    int len = string.length;
    next[0] = -1;
    int k = -1;
    for (int q = 1; q <= len - 1; q++) {
      while (k > -1 && string[k + 1] != string[q]) {
        k = next[k];
      }
      if (string[k + 1] == string[q]) {
        k = k + 1;
      }
      next[q] = k;
    }
  }

  List<TextRange> kmpSearchTextRange(String string) {
    List<TextRange> list = [];
    var m = this.length;
    var n = string.length;
    if (n > m) return list;
    List<int> next = List.filled(n, 0, growable: false);
    _calculateNext(string, next);
    int k = -1;
    for (int i = 0; i < m; i++) {
      while (k > -1 && string[k + 1] != this[i]) k = next[k];
      if (string[k + 1] == this[i]) k = k + 1;
      if (k == n - 1) {
        k = -1; //重新初始化，寻找下一个
        list.add(TextRange(start: i - n + 1, end:  i + 1));
      }
    }
    return list;
  }
}





lib/src/widget/string_parser.dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter_highlight_text/src/util/highlight_util.dart';
import 'package:string_scanner/string_scanner.dart';

class StringParser {
  final String content;
  final String splitString;
  final TextStyle style;
  final TextStyle highLightStyle;
  final WordClickAction? wordClickAction;
  late StringScanner _scanner;
  List<SpanBean> _spans = [];

  StringParser(
      {required this.content,
        required this.splitString,
        required this.style,
        required this.highLightStyle, this.wordClickAction});

  void parseContent() {
    while (!_scanner.isDone) {
      if (_scanner.scan(RegExp("$splitString.*?$splitString"))) {
        int startIndex = _scanner.lastMatch!.start;
        int endIndex = _scanner.lastMatch!.end;
        _spans.add(SpanBean(startIndex, endIndex));
      }
      if (!_scanner.isDone) {
        _scanner.position++;
      }
    }
  }

  InlineSpan parser() {
    _scanner = StringScanner(content);
    if (splitString.isEmpty) return TextSpan(style: style, children: [TextSpan(text: content, style: style)]);
    parseContent();
    final List<TextSpan> spans = <TextSpan>[];
    int currentPosition = 0;
    for (SpanBean span in _spans) {
      if (currentPosition != span.start)
        spans.add(
            TextSpan(text: content.substring(currentPosition, span.start), style: style));
      spans.add(TextSpan(
          style: highLightStyle,
          text: span.text(content),
          recognizer: TapGestureRecognizer()
            ..onTap = () {
              if (wordClickAction != null) wordClickAction!(span.text(content));
            }));
      currentPosition = span.end;
    }
    if (currentPosition != content.length)
      spans.add(TextSpan(
        text: content.substring(currentPosition, content.length),
        style: style
      ));
    return TextSpan(children: spans);
  }
}

class SpanBean {
  SpanBean(this.start, this.end);

  final int start;
  final int end;

  String text(String src) {
    return src.substring(start + 1, end - 1);
  }
}

lib/src/widget/hight_text_widget.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter_highlight_text/src/util/highlight_util.dart';
import 'package:flutter_highlight_text/src/util/kmp_util.dart';
import 'package:flutter_highlight_text/src/widget/string_parser.dart';

class HighLightText extends StatefulWidget {
  final String text;
  final TextStyle style;
  final HighLightType highLightType;
  final TextStyle? highLightTextStyle;
  final List<String>? highLightWords;
  final String? splitString;
  final WordClickAction? wordClickAction;

  const HighLightText(
      {Key? key,
      required this.text,
      required this.style,
      this.highLightTextStyle,
      required this.highLightType,
      this.highLightWords,
      this.splitString,
      this.wordClickAction})
      : super(key: key);

  @override
  State<StatefulWidget> createState() => _HighLightTextState();
}

class _HighLightTextState extends State<HighLightText> {
  late InlineSpan span;
  late StringParser parser;

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Text.rich(buildSpan());
  }

  InlineSpan buildSpan() {
    if (widget.highLightType == HighLightType.kerWords) {
      if (widget.highLightWords == null) {
        return TextSpan(children: [
          TextSpan(
            style: widget.style,
            text: widget.text,
          )
        ]);
      } else {
        return TextSpan(children: buildKeyWordsSpans());
      }
    } else {
      parser = StringParser(
        content: widget.text,
        splitString: widget.splitString ?? '',
        style: widget.style,
        highLightStyle: widget.highLightTextStyle ?? widget.style,
        wordClickAction: widget.wordClickAction,
      );
      span = parser.parser();
      return span;
    }
  }

  List<TextSpan> buildKeyWordsSpans() {
    List<TextRange> _ranges = [];
    List<TextSpan> spans = <TextSpan>[];
    widget.highLightWords!.forEach((element) {
      _ranges.addAll(KMP.kmpSearchTextRange(widget.text, element));
    });
    _ranges.sort((left, right) => left.start.compareTo(right.start));
    int currentPosition = 0;
    for (var i = 0; i < _ranges.length; i++) {
      TextRange range = _ranges[i];
      spans.add(TextSpan(
        style: widget.style,
        text: widget.text.substring(currentPosition, range.start),
      ));
      spans.add(TextSpan(
          text: widget.text.substring(range.start, range.end),
          style: widget.highLightTextStyle,
          recognizer: TapGestureRecognizer()
            ..onTap = () {
              if (widget.wordClickAction != null) widget.wordClickAction!(widget.text.substring(range.start, range.end));
            }));
      currentPosition = range.end;
    }
    return spans;
  }
}



lib/src/widget/highlight_text_widget.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter_highlight_text/src/util/highlight_util.dart';
import 'package:flutter_highlight_text/src/util/kmp_util.dart';
import 'package:flutter_highlight_text/src/widget/string_parser.dart';

class HighLightText extends StatefulWidget {
  final String text;
  final TextStyle style;
  final HighLightType highLightType;
  final TextStyle? highLightTextStyle;
  final List<String>? highLightWords;
  final String? splitString;
  final WordClickAction? wordClickAction;

  const HighLightText(
      {Key? key,
      required this.text,
      required this.style,
      this.highLightTextStyle,
      required this.highLightType,
      this.highLightWords,
      this.splitString,
      this.wordClickAction})
      : super(key: key);

  @override
  State<StatefulWidget> createState() => _HighLightTextState();
}

class _HighLightTextState extends State<HighLightText> {
  late InlineSpan span;
  late StringParser parser;

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Text.rich(buildSpan());
  }

  InlineSpan buildSpan() {
    if (widget.highLightType == HighLightType.keyWords) {
      if (widget.highLightWords == null) {
        return TextSpan(children: [
          TextSpan(
            style: widget.style,
            text: widget.text,
          )
        ]);
      } else {
        return TextSpan(children: buildKeyWordsSpans());
      }
    } else {
      parser = StringParser(
        content: widget.text,
        splitString: widget.splitString ?? '',
        style: widget.style,
        highLightStyle: widget.highLightTextStyle ?? widget.style,
        wordClickAction: widget.wordClickAction,
      );
      span = parser.parser();
      return span;
    }
  }

  List<TextSpan> buildKeyWordsSpans() {
    List<TextRange> _ranges = [];
    List<TextSpan> spans = <TextSpan>[];
    widget.highLightWords!.forEach((element) {
      _ranges.addAll(widget.text.kmpSearchTextRange(element));
    });
    _ranges.sort((left, right) => left.start.compareTo(right.start));
    int currentPosition = 0;
    for (var i = 0; i < _ranges.length; i++) {
      TextRange range = _ranges[i];
      spans.add(TextSpan(
        style: widget.style,
        text: widget.text.substring(currentPosition, range.start),
      ));
      spans.add(TextSpan(
          text: widget.text.substring(range.start, range.end),
          style: widget.highLightTextStyle,
          recognizer: TapGestureRecognizer()
            ..onTap = () {
              if (widget.wordClickAction != null) widget.wordClickAction!(widget.text.substring(range.start, range.end));
            }));
      currentPosition = range.end;
    }
    return spans;
  }
}



example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_highlight_text/flutter_highlight_text.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;
  String text =
      '可能说起 Flutter 绘制，大家第一反应就是用 `CustomPaint` 组件，自定义 `CustomPainter` 对象来画。Flutter 中所有可以看得到的组件，比如 Text、Image、Switch、Slider 等等，追其根源都是`画出来`的，但通过查看源码可以发现，Flutter 中绝大多数组件并不是使用 `CustomPaint` 组件来画的，其实 `CustomPaint` 组件是对框架底层绘制的一层封装。这个系列便是对 Flutter 绘制的探索，通过`测试`、`调试`及`源码分析`来给出一些在绘制时`被忽略`或`从未知晓`的东西，而有些要点如果被忽略，就很可能出现问题。';

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug painting" (press "p" in the console, choose the
          // "Toggle Debug Paint" action from the Flutter Inspector in Android
          // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
          // to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            HighLightText(
              style: TextStyle(fontSize: 14, color: Colors.black),
              highLightTextStyle:
              TextStyle(color: Colors.red, fontSize: 18, fontWeight: FontWeight.bold),
              text: text,
              highLightWords: ['Flutter', '可'],
              splitString: '`',
              highLightType: HighLightType.keyWords,
              wordClickAction: (word) {
                print("=====$word");
              },
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
